dart/common/detail/Addon.h

```diff
@@ -0,0 +1,338 @@
+  if(nullptr == castedManager)\
+  {\
+    dterr << "[" << typeid(Base).name() << "::" << #func << "] Attempting to "\
+          << "use a [" << typeid(newManager).name() << "] type manager, but "\
+          << "this Addon is only designed to be attached to a ["\
+          << typeid(ManagerType).name() << "] type manager. This may cause "\
+          << "undefined behavior!\n";\
+    assert(false);\
+  }
+
+#define DART_COMMON_CAST_NEW_MANAGER_TYPE_AND_RETURN_NULL_IF_BAD(\
+  Base, ManagerType, newManager, castedManager, func)\
+  DART_COMMON_CAST_NEW_MANAGER_TYPE(Base, ManagerType, newManager, castedManager, func)\
+  if(nullptr == castedManager) return nullptr;
+
+//==============================================================================
+template <class Base, typename StateData,
+          class ManagerType, void (*updateState)(Base*)>
+AddonWithProtectedState<Base, StateData, ManagerType, updateState>::
+AddonWithProtectedState(
+    AddonManager* mgr, const StateData& state)
+  : Addon(mgr),
+    mState(state)
+{
+  // Do nothing
+}
+
+//==============================================================================
+template <class Base, typename StateData,
+          class ManagerType, void (*updateState)(Base*)>
+void AddonWithProtectedState<Base, StateData, ManagerType, updateState>::
+setAddonState(const Addon::State& otherState)
+{
+  setState(static_cast<const State&>(otherState));
+}
+
+//==============================================================================
+template <class Base, typename StateData,
+          class ManagerType, void (*updateState)(Base*)>
+const Addon::State*
+AddonWithProtectedState<Base, StateData, ManagerType, updateState>::
+getAddonState() const
+{
+  return &mState;
+}
+
+//==============================================================================
+template <class Base, typename StateData,
+          class ManagerType, void (*updateState)(Base*)>
+void AddonWithProtectedState<Base, StateData, ManagerType, updateState>::
+setState(const StateData& state)
+{
+  static_cast<StateData&>(mState) = state;
+  UpdateState(static_cast<Base*>(this));
+}
+
+//==============================================================================
+template <class Base, typename StateData,
+          class ManagerType, void (*updateState)(Base*)>
+auto AddonWithProtectedState<Base, StateData, ManagerType, updateState>::
+getState() const -> const State&
+{
+  return mState;
+}
+
+//==============================================================================
+template <class Base, typename StateData,
+          class ManagerType, void (*updateState)(Base*)>
+std::unique_ptr<Addon>
+AddonWithProtectedState<Base, StateData, ManagerType, updateState>::
+    cloneAddon(AddonManager* newManager) const
+{
+  DART_COMMON_CAST_NEW_MANAGER_TYPE_AND_RETURN_NULL_IF_BAD(
+        Base, ManagerType, newManager, castedManager, clone);
+  return std::unique_ptr<Base>(new Base(newManager, mState));
```

Comment:
The constructor is supposed to expect a generic ``AddonManager*``, so that's what we give it. If we gave it the casted version, there is a risk that it will be a plain ``nullptr``, which is not a scenario that I think the derived Addon type should have to handle.